### PR TITLE
LLMServerPool Implementation

### PR DESCRIPTION
**Pull request by kfswain**
Implementation to integrate LLMServerPool and Pod reconciliation into the I-GW EndpointPicker(ext-proc)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,12 @@
 ## Multistage build
-FROM golang:1.22.5-alpine as build
+FROM golang:1.23-alpine as build
 ENV CGO_ENABLED=0
 ENV GOOS=linux
 ENV GOARCH=amd64
 
 WORKDIR /src
 COPY . .
+WORKDIR /src/pkg/ext-proc
```
Comment (kfswain) [inline Dockerfile:9]: Temp. I'll move this dockerfile back into the ext-proc dir when I'm finished. Just needed the root context to see all local packages (the API)

File: pkg/ext-proc/backend/pod_controller.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+func NewPodController(
+	ctx context.Context,
+	llmServerPoolName string,
+	datastore *Datastore,
+	kubeclientset kubernetes.Interface,
+	podInformer coreinformers.PodInformer) *PodController {
+
+	logger := klog.FromContext(ctx)
+	utilruntime.Must(scheme.AddToScheme(scheme.Scheme))
+	logger.V(4).Info("Creating event broadcaster")
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:54]: this is printed once every restart, probably worth upping the level

File: pkg/ext-proc/backend/pod_controller.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+	controllerName := controllerNamePrefix + llmServerPoolName + "-pod-controller"
+	eventBroadcaster := record.NewBroadcaster(record.WithContext(ctx))
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerName})
+	ratelimiter := workqueue.NewTypedMaxOfRateLimiter(
+		workqueue.NewTypedItemExponentialFailureRateLimiter[cache.ObjectName](5*time.Millisecond, 1000*time.Second),
+		&workqueue.TypedBucketRateLimiter[cache.ObjectName]{Limiter: rate.NewLimiter(rate.Limit(50), 300)},
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:63]: Can you please add a comment explaining in layman's terms the effect of this configuration? also, if this was a recommendation made somewhere (in api-machinary), then it is worth citing it so we go back to it if a new recommendation was made.

File: pkg/ext-proc/backend/pod_controller.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+	// Wait for the caches to be synced before starting workers
+	logger.Info("Waiting for informer caches to sync")
+
+	if ok := cache.WaitForCacheSync(ctx.Done(), c.podsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	logger.Info("Starting workers", "count", workers)
+	// Launch two workers to process Foo resources
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:102]: update comment

File: pkg/ext-proc/backend/pod_controller.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+			}
+		}
+	}
+}
+
+func (c *PodController) updateDatastore(objName cache.ObjectName) error {
+	pod, err := c.podLister.Pods(objName.Namespace).Get(objName.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			//Delete pod from datastore in a threadsafe way
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:185]: it doesn't look like we are doing this.

File: pkg/ext-proc/backend/pod_controller.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+			return nil
+		}
+		return err
+	}
+
+	if !c.datastore.LabelsMatch(pod.ObjectMeta.Labels) {
+		c.datastore.Pods.Delete(pod.Name)
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:192]: we should also delete it from the store if the pod is marked for deletion (deletionTimestamp is set)

**Review by ahg-g — COMMENTED** [review]
We should use controller-runtime instead of creating and managing workqueues and informerFactory etc.  https://github.com/kubernetes-sigs/controller-runtime

See the example here, it is simpler: https://github.com/kubernetes-sigs/controller-runtime/blob/main/examples/crd/main.go

File: examples/poc/ext-proc/go.mod [hunk 7]
```diff
@@ -1,4 +1,4 @@
-module ext-proc
+module ext-proc-poc
```
[the remark this answers is not in the document]
Comment (kfswain) [reply examples/poc/ext-proc/go.mod:1]: Go.work cant have 2 modules of the same name

File: pkg/ext-proc/backend/pod_controller.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+type PodController struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Record    record.EventRecorder
+	Datastore *K8sDatastore
+}
+
+func (c *PodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:22]: We need to check health status.

I actually suggest that the LLMServerPool reconciler to create a Service, which will create EndpoinSlices, which will include the health/readiness status.

With this, we don't need a pod reconciler; we will need an EndpointSlice reconciler to update the data store with the healthy pods only. wdyt?

Comment (ahg-g) [reply pkg/ext-proc/backend/pod_controller.go:22]: The other option is to keep this pod reconciler, but check the readiness status of the pod remove unhealthy pods from the store

Comment (kfswain) [reply pkg/ext-proc/backend/pod_controller.go:22]: I think I'll keep this PR as a pod reconciler. We can migrate to EndpointSlice later if we want, but I believe the only gain would be the management of the status/health of pods for us. Right?

Comment (ahg-g) [reply pkg/ext-proc/backend/pod_controller.go:22]: Discussed offline, working with EndpointSlices offers more info on the endpoints in one place, including health status, locality information etc.

File: pkg/ext-proc/backend/llmserverpool_controller.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+	klog.V(1).Info("reconciling LLMServerPool", req.NamespacedName)
+
+	serverPool := &v1alpha1.LLMServerPool{}
+	if err := c.Get(ctx, req.NamespacedName, serverPool); err != nil {
+		klog.Error(err, "unable to get LLMServerPool")
+		return ctrl.Result{}, err
+	}
+
+	needNewService := true
+	if needNewService {
```
Comment (ahg-g) [inline pkg/ext-proc/backend/llmserverpool_controller.go:47]: flag here is always true

File: pkg/ext-proc/backend/llmserverpool_controller.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+					{
+						Port: int32(c.Port),
+					},
+				},
+				Selector: labels,
+			},
+		}
+		if err := c.Create(ctx, service); err != nil {
+			klog.Error(err, "unable to create service")
+			return ctrl.Result{RequeueAfter: time.Second * 5}, err
```
Comment (ahg-g) [inline pkg/ext-proc/backend/llmserverpool_controller.go:68]: do we need to behave differently if the error indicates that the service exists already?

File: pkg/ext-proc/backend/llmserverpool_controller.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+	serverPool := &v1alpha1.LLMServerPool{}
+	if err := c.Get(ctx, req.NamespacedName, serverPool); err != nil {
+		klog.Error(err, "unable to get LLMServerPool")
+		return ctrl.Result{}, err
+	}
+
+	needNewService := true
+	if needNewService {
+		labels, err := metav1.LabelSelectorAsMap(&c.Datastore.LLMServerPool.Spec.ModelServerSelector)
```
Comment (ahg-g) [inline pkg/ext-proc/backend/llmserverpool_controller.go:48]: don't we want to change the `ModelServerSelector` to a map?

File: pkg/ext-proc/backend/pod_controller.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+)
+
+type PodController struct {
```
Comment (ahg-g) [inline pkg/ext-proc/backend/pod_controller.go:15]: why do we still need this? The EndpointSlice reconciler will be the one looking up the endpoints and storing them in the datastore

File: pkg/ext-proc/backend/endpointslice_controller.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+	c.updateDatastore(endpointSlice)
+
+	return ctrl.Result{}, nil
+}
+
+func (c *EndpointSliceController) ownsEndPointSlice(fullName types.NamespacedName) bool {
+	name := c.ServerPoolName
+	if c.Zone != "" {
+		name = c.ServerPoolName + "-" + c.Zone
```
Comment (ahg-g) [inline pkg/ext-proc/backend/endpointslice_controller.go:48]: can we please document the assumption we are making here regarding the name?

**Review by ahg-g — COMMENTED** [review]
I started to review this, but it seems this is still in development, right?

**Comment by ahg-g** [comment]
/hold

just so this doesn't get merged by mistake

**Comment by kfswain** [comment]
> I started to review this, but it seems this is still in development, right?

Yes, I should have put a hold on it (ty for that). I'll take the hold off when it's complete.

File: pkg/ext-proc/backend/datastore.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+type K8sDatastore struct {
+	LLMServerPool *v1alpha1.LLMServerPool
+	Pods          *sync.Map
+	Port          string
+}
+
+func (ds *K8sDatastore) GetPodIPs() []string {
+	var ips []string
+	ds.Pods.Range(func(name, pod any) bool {
+
```
Comment (ahg-g) [inline pkg/ext-proc/backend/datastore.go:23]: nit: empty line

File: pkg/ext-proc/backend/datastore.go [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+	if err != nil {
+		klog.Error(err.Error())
+		return false
+	}
+	if selector == nil {
+		return false
+	}
+	set := labels.Set(podLabels)
+	return selector.Matches(set)
+
```
Comment (ahg-g) [inline pkg/ext-proc/backend/datastore.go:41]: nit: empty line

File: pkg/ext-proc/backend/datastore.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+	var ips []string
+	ds.Pods.Range(func(name, pod any) bool {
+
+		ips = append(ips, pod.(*corev1.Pod).Status.PodIP)
+		return true
+	})
+	return ips
+}
+
+func (ds *K8sDatastore) LabelsMatch(podLabels map[string]string) bool {
```
Comment (ahg-g) [inline pkg/ext-proc/backend/datastore.go:30]: do we still need this?

File: pkg/ext-proc/backend/llmserverpool_controller.go [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+	Zone           string
+}
+
+func (c *LLMServerPoolController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if req.NamespacedName.Name != c.ServerPoolName && req.NamespacedName.Namespace != c.Namespace {
+		return ctrl.Result{}, nil
+	}
+	klog.V(1).Info("reconciling LLMServerPool", req.NamespacedName)
+
+	serverPool := &v1alpha1.LLMServerPool{}
```
Comment (ahg-g) [inline pkg/ext-proc/backend/llmserverpool_controller.go:37]: aren't we supposed to create a Service here?

Comment (kfswain) [reply pkg/ext-proc/backend/llmserverpool_controller.go:37]: I realized that it would be confusing to have the EPP reconcile and create an object on the LLMServerPool's behalf, in addition to whatever controller created the EPP. Instead, the implementation of the LLMServerPool controller should create the service and pass the service name in on startup. I think centralizing the object creation/control will make for a cleaner interface/experience

File: pkg/ext-proc/backend/datastore.go (hunk 16 above)
Comment (kfswain) [reply pkg/ext-proc/backend/datastore.go]: Nope! Deleting

File: pkg/ext-proc/backend/llmserverpool_controller.go (hunk 17 above)
Comment (kfswain) [reply pkg/ext-proc/backend/llmserverpool_controller.go]: Leaving this comment unresolved for posterity. But discussed further offline. And it sounds like we are in agreement here
